coffea/processor/executor.py

```diff
@@ -99,16 +142,136 @@ def futures_executor(items, function, accumulator, workers=1, status=True, unit=
+            Leave zero (default) for no compression.
+    """
+    client = kwargs.pop('client')
+    ntree = kwargs.pop('treereduction', 20)
+    status = kwargs.pop('status', True)
+    clevel = kwargs.pop('compression', 0)
+    reducer = _reduce
+    if clevel > 0:
+        function = _compression_wrapper(clevel, function)
+        reducer = _compression_wrapper(clevel, reducer)
+
+    futures = client.map(function, items)
+    while len(futures) > 1:
+        futures = client.map(
+            reducer,
+            [futures[i:i + ntree] for i in range(0, len(futures), ntree)],
+            priority=1,
+        )
+    if status:
+        from dask.distributed import progress
+        progress(futures, multi=False, notebook=False)
+    accumulator += _maybe_decompress(futures.pop().result())
     return accumulator
 
 
-def _work_function(item, flatten=False, savemetrics=False, mmap=False, **_):
-    dataset, fn, treename, chunksize, index, processor_instance = item
+def parsl_executor(items, function, accumulator, **kwargs):
+    """Execute using parsl pyapp wrapper
+
+    Parameters
+    ----------
+        items : list
+            List of input arguments
+        function : function
+            A function to be called on each input, which returns an accumulator instance
+        accumulator : AccumulatorABC
+            An accumulator to collect the output of the function
+        config : parsl.config.Config, optional
+            A parsl DataFlow configuration object. Necessary if there is no active kernel
+
+            .. note:: In general, it is safer to construct the DFK with ``parsl.load(config)`` prior to calling this function
+        status : bool
+            If true (default), enable progress bar
+        unit : str
+            Label of progress bar unit
+        desc : str
+            Label of progress bar description
+        compression : int, optional
+            Compress accumulator outputs in flight with LZ4, at level specified.
+            Leave zero (default) for no compression.
+    """
+    import parsl
+    from parsl.app.app import python_app
+    from .parsl.timeout import timeout
+    status = kwargs.pop('status', True)
+    unit = kwargs.pop('unit', 'items')
+    desc = kwargs.pop('desc', 'Processing')
+    clevel = kwargs.pop('compression', 0)
+    if clevel > 0:
+        function = _compression_wrapper(clevel, function)
+    add_fn = _iadd
+
+    cleanup = False
+    config = kwargs.pop('config', None)
+    try:
+        parsl.dfk()
+    except RuntimeError:
+        cleanup = True
+        pass
+    if cleanup and config is None:
+        raise RuntimeError("No active parsl DataFlowKernel, must specify a config to construct one")
+    elif not cleanup and config is not None:
+        raise RuntimeError("An active parsl DataFlowKernel already exists")
+    else:
+        parsl.clear()
+        parsl.load(config)
+
+    # wrap with timeout() ?
+    app = python_app(function)
```

Comment:
@lgray I noticed what appears to be a watchdog timer built into parsl now.  Do you know if that is new, or it was there and didn't work?

Reply:
It may still not work for python apps. In 0.8.0 it definitely doesn't.
@annawoodard what's the status in parsl master?

Reply:
@lgray @nsmith: The [parsl timeout](https://github.com/Parsl/parsl/blob/master/parsl/app/python.py#L14-L32) works by starting a timer and injecting an exception into the current thread if the timer goes off. This has the advantage that it doesn't kill the Parsl worker, but it has the disadvantage that it may not work if the hang is happening outside Python (for example, in some underlying C++ code XrootD is running). I recommend you use the timeout implemented [here](https://github.com/CoffeaTeam/coffea/blob/master/coffea/processor/parsl/timeout.py) which uses a SIGALRM to blast the whole worker. On master, I've added the machinery to transparently restart the worker. As long as you have `retries > 0` in your Parsl config, any timed-out tasks will automatically be resubmitted. I've tested this works on XrootD hangs.